BraveWallet/Crypto/Stores/SwapTokenStore.swift

```diff
@@ -32,48 +33,384 @@ public class SwapTokenStore: ObservableObject {
   private let rpcController: BraveWalletEthJsonRpcController
   private let assetRatioController: BraveWalletAssetRatioController
   private let swapController: BraveWalletSwapController
+  private let transactionController: BraveWalletEthTxController
   private var accountInfo: BraveWallet.AccountInfo?
+  private var slippage = 0.005 {
+    didSet {
+      timer?.invalidate()
+      timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: false, block: { [weak self] _ in
+        self?.fetchPriceQuote(base: .perSellAsset)
+      })
+    }
+  }
+  private var updatingPriceQuote = false
+  private var timer: Timer?
+  
+  enum SwapParamsBase {
+    // calculating based on sell asset amount
+    case perSellAsset
+    // calculating based on buy asset amount
+    case perBuyAsset
+  }
+  
+  enum SwapState {
+    // when there is an error occurs. associated with an error message
+    case error(String)
+    // when need to activate bigger allowance. associated with a spender address
+    case activateAllowance(String)
+    // when able to swap
+    case swap
+    // has not passed validation
+    case idle
+  }
   
   public init(
+    keyringController: BraveWalletKeyringController,
     tokenRegistry: BraveWalletERCTokenRegistry,
     rpcController: BraveWalletEthJsonRpcController,
     assetRatioController: BraveWalletAssetRatioController,
-    swapController: BraveWalletSwapController
+    swapController: BraveWalletSwapController,
+    transactionController: BraveWalletEthTxController
   ) {
+    self.keyringController = keyringController
     self.tokenRegistry = tokenRegistry
     self.rpcController = rpcController
     self.assetRatioController = assetRatioController
     self.swapController = swapController
+    self.transactionController = transactionController
+    
+    self.keyringController.add(self)
+    self.rpcController.add(self)
   }
   
   private func fetchTokenBalance(
     for token: BraveWallet.ERCToken,
-    completion: @escaping (_ balance: Double?) -> Void
+    completion: @escaping (_ balance: BDouble?) -> Void
   ) {
     guard let account = accountInfo else {
       completion(nil)
       return
     }
     
     rpcController.balance(for: token, in: account) { balance in
-      completion(balance)
+      if let value = balance {
+        completion(BDouble(value))
+      }
+    }
+  }
+  
+  private func stringToBytes(_ string: String) -> [NSNumber] {
+    let clean = string.removingHexPrefix
+    
+    var result = [NSNumber](repeating: 0, count: string.count / 2)
+    for i in stride(from: 0, to: clean.count, by: 2) {
+        let start = clean.index(clean.startIndex, offsetBy: i)
+        let end = clean.index(clean.startIndex, offsetBy: i+2)
+        let range = start..<end
+        let mysubstring = clean[range]
+        if let value = UInt8(mysubstring, radix: 16) {
+            result[i/2] = NSNumber(value: value)
+        }
+    }
+    
+    return result
+  }
```

Comment:
This will break if an odd count string is delivered (not sure if this is guaranteed not to happen from ETH RPC calls), but to be safe we should probably insert a leading 0 if its odd-numbered. Also since we already include `swift-algorithms` as a library we can utilize the `chunked(ofCount:)` API to simplify this:

```swift
extension Array where Element == NSNumber {
  init?(hexString: String) {
    var value = hexString
    if value.count % 2 != 0 {
      value.insert("0", at: value.startIndex)
    }
    if !value.allSatisfy(\.isHexDigit) {
      return nil
    }
    self.init(
      value.chunks(ofCount: 2)
        .compactMap { UInt8($0, radix: 16) }
        .map(NSNumber.init(value:))
    )
  }
}
```

Something like the above is a bit safer, and we could unit test if its placed in a separate file